kotlin/display-kml/src/main/java/com/esri/arcgisruntime/sample/displaykml/MainActivity.kt

```diff
@@ -0,0 +1,169 @@
+        // create a KML layer created from the KML data set and display it on the map
+        val kmlLayer = KmlLayer(kmlDataset)
+        display(kmlLayer)
+        // report layers if failed to load
+        reportErrors(kmlLayer, "Failed to load KML layer from URL: ")
+    }
+
+    /**
+     * Display a KML layer from a portal item.
+     */
+    fun changeSourceToPortalItem() {
+        // create a portal to ArcGIS Online
+        val portal = Portal(getString(R.string.arcgis_online_url))
+        // create a portal item from a KML item id
+        val portalItem = PortalItem(portal, getString(R.string.kml_item_id))
+
+        // a KML layer created from an ArcGIS Online portal item
+        val kmlLayer = KmlLayer(portalItem)
+        display(kmlLayer)
+        // report layers if failed to load
+        reportErrors(kmlLayer, "Failed to load KML layer from portal item: ")
+    }
+
+    /**
+     * Display a kml layer from external storage.
+     */
+    fun changeSourceToFileExternalStorage() {
+        // get the data set stored locally in device external storage
+        val file = getExternalFilesDir(null)?.path + getString(R.string.kml_path)
+        val kmlDataset = KmlDataset(file)
+        // create a KML layer from the locally stored data set
+        val kmlLayer = KmlLayer(kmlDataset)
+        display(kmlLayer)
+        // report layers if failed to load
+        reportErrors(kmlLayer, "Failed to load kml data set from external storage: " + file)
+    }
+
+    /**
+     * Reports any loading errors of the KML datasets
+     *
+     * @param kmlData any loadable containing a kml data set e.g. KmlDataset or KmlLayer
+     * @param string the error message to display
+     */
+    fun reportErrors(kmlData: Loadable, string: String) {
+        // report errors if failed to load
+        kmlData.addDoneLoadingListener {
+            if (kmlData.loadStatus != LoadStatus.LOADED) {
+                val error = string + kmlData.loadError.message
+                Toast.makeText(this, error, Toast.LENGTH_LONG).show()
+                Log.e(TAG, error)
+            }
+        }
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        menuInflater.inflate(R.menu.kml_sources, menu)
+        return super.onCreateOptionsMenu(menu)
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        when (item.itemId) {
+            R.id.kmlFromUrl -> changeSourceToURL()
+            R.id.kmlFromPortal -> changeSourceToPortalItem()
+            R.id.kmlFromExternalStorage -> changeSourceToFileExternalStorage()
+            else -> return super.onOptionsItemSelected(item)
+        }
+        return true
+    }
+
+    override fun onPause() {
+        super.onPause()
+        mapView.pause()
+    }
+
+    override fun onResume() {
+        super.onResume()
+        mapView.resume()
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        mapView.dispose()
+    }
```

Comment:
override fun onPause() {
        mapView.pause()
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
        mapView.resume()
    }

    override fun onDestroy() {
        mapView.dispose()
        super.onDestroy()
    }

Your code is valid, but we try to order it like this as a subtle reminder to the Android lifecycle:

https://devtopia.esri.com/runtime/common-samples/wiki/Android-Samples-Repo%3A-A-Contribution-Guide#handle-life-cycles